packages/stacking/src/utils.ts

```diff
@@ -23,68 +37,99 @@ export class InvalidAddressError extends Error {
   }
 }
 
-export const BitcoinNetworkVersion = {
-  mainnet: {
-    P2PKH: 0x00, // 0
-    P2SH: 0x05, // 5
-  },
-  testnet: {
-    P2PKH: 0x6f, // 111
-    P2SH: 0xc4, // 196
-  },
-} as const;
-
-export function btcAddressVersionToHashMode(btcAddressVersion: number): AddressHashMode {
+/** @ignore */
+export function btcAddressVersionToLegacyHashMode(btcAddressVersion: number): PoXAddressVersion {
   switch (btcAddressVersion) {
     case BitcoinNetworkVersion.mainnet.P2PKH:
-      return AddressHashMode.SerializeP2PKH;
+      return PoXAddressVersion.P2PKH;
     case BitcoinNetworkVersion.testnet.P2PKH:
-      return AddressHashMode.SerializeP2PKH;
+      return PoXAddressVersion.P2PKH;
     case BitcoinNetworkVersion.mainnet.P2SH:
-      return AddressHashMode.SerializeP2SH;
+      return PoXAddressVersion.P2SH;
     case BitcoinNetworkVersion.testnet.P2SH:
-      return AddressHashMode.SerializeP2SH;
+      return PoXAddressVersion.P2SH;
     default:
       throw new Error('Invalid pox address version');
   }
 }
 
-export function hashModeToBtcAddressVersion(
-  hashMode: AddressHashMode,
-  network: 'mainnet' | 'testnet'
-): number {
-  if (!['mainnet', 'testnet'].includes(network)) {
-    throw new Error(`Invalid network argument: ${network}`);
-  }
-  switch (hashMode) {
-    case AddressHashMode.SerializeP2PKH:
-      return BitcoinNetworkVersion[network].P2PKH;
-    case AddressHashMode.SerializeP2SH:
-      return BitcoinNetworkVersion[network].P2SH;
-    default:
-      throw new Error(`Invalid pox address hash mode: ${hashMode}`);
-  }
+/** @ignore */
+function nativeAddressToSegwitVersion(
+  witnessVersion: number,
+  dataLength: number
+): PoXAddressVersion {
+  if (witnessVersion === SEGWIT_V0 && dataLength === 20) return PoXAddressVersion.P2WPKH;
+  if (witnessVersion === SEGWIT_V0 && dataLength === 32) return PoXAddressVersion.P2WSH;
+  if (witnessVersion === SEGWIT_V1 && dataLength === 32) return PoXAddressVersion.P2TR;
+
+  throw new Error(
+    'Invalid native segwit witness version and byte length. Currently, only P2WPKH, P2WSH, and P2TR are supported.'
+  );
+}
+
+function bech32Decode(btcAddress: string) {
+  const { words: bech32Words } = bech32.decode(btcAddress);
+  const witnessVersion = bech32Words[0];
+
+  if (witnessVersion > 0)
+    throw new Error('Addresses with a witness version >= 1 should be encoded in bech32m');
+
+  return {
+    witnessVersion,
+    data: bech32.fromWords(bech32Words.slice(1)),
+  };
 }
 
-export function getAddressHashMode(btcAddress: string) {
+function bech32MDecode(btcAddress: string) {
+  const { words: bech32MWords } = bech32m.decode(btcAddress);
+  const witnessVersion = bech32MWords[0];
+
+  if (witnessVersion == 0)
+    throw new Error('Addresses with witness version 1 should be encoded in bech32');
+
+  return {
+    witnessVersion,
+    data: bech32m.fromWords(bech32MWords.slice(1)),
+  };
+}
+
+function decodeNativeSegwitBtcAddress(btcAddress: string): {
```

Comment:
Nit: the mildly inefficient (and mildly annoying in debuggers) try/catch here can be replaced with something like the following:
https://github.com/hirosystems/stacks-blockchain-api/blob/ae0eb65c4d6901db172f2b4ea751817d8ef8c05c/src/pox-helpers.ts#L180-L189